### ISSUE TITLE
[BUG] IAM roles are not passed on to MPI Launcher and worker pods

### ISSUE THREAD

**Issue by bimtauer** [flyteorg/flyte#3043]
### Describe the bug

An IAM role defined in the launch dialog, in a launchplan or in the k8s defaults will not be passed on to an MPI Job's launcher and worker pods and thus results in S3 errors if iam role authentication is used. 

This also concerns the default IAM role if declared which according to my understanding is usually applied together with other defaults as part of [AddObjectMetadata](https://github.com/flyteorg/flytepropeller/blob/4d8f800f508554534f91b9dcc474d4c6b6a1f874/pkg/controller/nodes/task/k8s/plugin_manager.go#L104).

Good news is all annotations are properly set on the top level `MPIJob` manifest - however they are not set on the P[odTemplates for the worker and launcher pods withing that generated manifest](https://github.com/flyteorg/flyteplugins/blob/cdfc191012c2b2d5e3d956dec031df790b75386e/go/tasks/plugins/k8s/kfoperators/mpi/mpi.go#L89-L101).

### Expected behavior

A default IAM role (defined as a default annotation) or an IAM role provided in the launch dialog should be passed down to an MPI Jobs worker and launcher pods

### Additional context to reproduce

Run a workflow with an MPI task and specify a custom iam role upon execution. 

### Screenshots

![image](https://user-images.githubusercontent.com/30375389/199288892-7274d2b0-a5a6-4fd9-9ef8-cf8a0f9bef03.png)

### Are you sure this issue hasn't been raised already?

- [X] Yes

### Have you read the Code of Conduct?

- [X] Yes

**Comment by hamersaw** [comment]
@bimtauer thanks for filing this issue! It looks like this is not just and MPI job issue but should probably be propagated through to the pytorch and tensorflow plugins as well.

So the `AddObjectMetadata` you mentioned is only meant to be executed on the top-level k8s resource as you observed. Do you expect it to add labels / annotations / etc to the underlying worker pods (etc) as well?

As far as adding a service account this should be a very easy fix. We are already doing this in the [ray](https://github.com/flyteorg/flyteplugins/blob/cdfc191012c2b2d5e3d956dec031df790b75386e/go/tasks/plugins/k8s/ray/ray.go#L139-L144) and [spark](https://github.com/flyteorg/flyteplugins/blob/cdfc191012c2b2d5e3d956dec031df790b75386e/go/tasks/plugins/k8s/spark/spark.go#L95-L109) tasks. You should be able to call the `flytek8s.GetServiceAccountNameFromTaskExecutionMetadata` function to retrieve the correct account.

Is this something you might be able to contribute? I can offer a speedy review :)

**Comment by bimtauer** [comment]
@hamersaw thanks for looking into this. And yes we expect to run into the same issue with Spark jobs or any other complex manifest rendered by Flyte that itself declares own PodTemplates.

My first thought went back to `AddObjectmetadata`  since I in the past once submitted a proposal for implementing some templating logic there - I think since then you have created a more powerful templating logic [here](https://github.com/flyteorg/flyteplugins/pull/247). But I guess both of those paths are "top level" in the sense that they only affect k8s pods directly described by Flyte. 

Service accounts are definitely needed as well - we use those to authenticate pods with Vault. But afaik kube2iam relies solely on a matching annotation, so for aws access we'd need those to be propagated down as well. 

I suppose [this here](https://github.com/flyteorg/flyteplugins/blob/d95b86b4ba9e6fcab18b3ea834047c4248cf314a/go/tasks/plugins/k8s/kfoperators/mpi/mpi.go#L91-L98) is where the MPI spec is first put together. Maybe you can use your `flytek8s.BuildPodWithSpec(podTemplate, podSpec, primaryContainerName)` logic directly there (and in other plugins) to create the pod templates inside those plugin specific manifests?

**Comment by bimtauer** [comment]
@hamersaw Thanks for this summary, this sounds very much like what we need. 

One small point that still confuses me is when you would recommend using [PodTemplates](https://docs.flyte.org/en/latest/deployment/cluster_config/general.html#using-default-k8s-podtemplates) VS using the [k8s plugin config fields](https://github.com/flyteorg/flyteplugins/blob/902b902fcf487f30ebb5dbeee3bb14e17eb0ec21/go/tasks/pluginmachinery/flytek8s/config/config.go#L76). 

My understanding is that I can use PodTemplates to do anything I could have otherwise done in the k8s plugin config.yaml (i.e default  annotations, labels, tolerations etc.) and much more - I can create PodTemplates namespace-specific which means I can create different defaults for each project and domain. 

If the above is right then I would prefer only using PodTemplates in the future, so no need for point 2. at least from us.

**Comment by hamersaw** [comment]
@bimtauer so the default PodTemplate work was done because we believe it is a much better mechanism than exposing configuration through the k8s plugin. So you are 100% correct, we are hoping users slowly transition from k8s plugin config fields to entirely using PodTemplates.

**Comment by hamersaw** [comment]
Fixed by https://github.com/flyteorg/flyteplugins/pull/297
